app/src/dex/router/fill_route2.rs

```diff
@@ -0,0 +1,412 @@
+            // continue rolling over dust positions without allowing arbitrarily
+            // higher prices than the requested spill price?
+            if spill_price.is_some() && current_effective_price > spill_price {
+                tracing::debug!(
+                    // HACK: render infinite price as 0
+                    spill_price = %spill_price.unwrap_or_default(),
+                    current_effective_price = %current_effective_price.unwrap_or_default(),
+                    "exceeded spill price, breaking loop"
+                );
+                break 'filling;
+            }
+
+            // Find the next position to replace the one we just consumed.
+            'next_position: loop {
+                let next_position_id = match positions_by_price
+                    .get_mut(&pairs[constraining_index])
+                    .unwrap()
+                    .as_mut()
+                    .next()
+                    .await
+                    .transpose()?
+                {
+                    // If none is available, we can't keep filling, and need to break.
+                    None => {
+                        tracing::debug!(
+                            pair = ?pairs[constraining_index],
+                            "no more positions available for pair, breaking loop"
+                        );
+                        break 'filling;
+                    }
+                    // Otherwise, we need to check that the position is not already
+                    // part of the current frontier.
+                    Some(position_id) if !frontier_position_ids.contains(&position_id) => {
+                        position_id
+                    }
+                    // Otherwise, continue to the next position in the stream.
+                    Some(position_id) => {
+                        tracing::debug!(?position_id, "skipping position already in frontier");
+                        continue 'next_position;
+                    }
+                };
+
+                let next_position = self
+                    .position_by_id(&next_position_id)
+                    .await?
+                    .expect("indexed position should exist");
+
+                tracing::debug!(
+                    constraining_index,
+                    ?next_position_id,
+                    ?next_position,
+                    "replacing constraining position in frontier",
+                );
+
+                frontier_position_ids.insert(next_position_id);
+                frontier[constraining_index] = next_position;
+                // TODO: should we remove the old position's ID from the ID set?
+
+                // Having updated the frontier, it's time to break out of the loop.
+                break 'next_position;
+            }
+        }
+
+        // We need to save these positions, because we mutated their state, even
+        // if we didn't fully consume their reserves.
+        for position in frontier {
```

Comment:
I think this should be in the filling loop, otherwise we might execute against stale orders on later fils?